### DISCUSSION TITLE
HTTP proxy

### DISCUSSION THREAD

**Discussion by RafaelLinux** [rustdesk/rustdesk#5687] [Q&A] [answered]
In my intranet, we require to use an http(s) proxy. RustDesk nowadays only works with socks proxy, so it shows "Connecting to the RustDesk network" forever and I don't get an ID.

Please, could you add the HTTP(S) proxy feature?

Thank you

**Comment by RafaelLinux** [comment]
For reasons I don't imagine, what I'm asking for is not considered an enhancement and this issue is moved to a discussion https://github.com/rustdesk/rustdesk/discussions/831 so I'll close it.

**Comment by rustdesk** [comment]
ok, as more and more require this feature, i will put in milestone #918, though it is alreay very long.

**Comment by RafaelLinux** [comment]
Thanks!!!!!

**Comment by bit-ryder** [comment]
a http proxy feature would be xtra great, unfortunately i can't program in rust. ;-))
thanks for the great software!

greetings

**Comment by michag86** [comment]
To make it proxy-capable, the entire data traffic would have to be tunneled via https port 443. The most proxies restrict connections to common ports. The new listener for 443 should also be publishable via reverse proxy.

**Comment by yuluo-zy** [comment]
I think this function, I can realize it, can you provide some guidance and assistance,where should I complete this function in the project. @rustdesk

**Reply by AlexanderGabriel** [reply]
Please tell me, if or how i can support you. Can not code in rust but can tell you what we need and test. Maybe ask my boss if we can spend money to have this feature and then kick out our proprietary solution.

**Reply by yuluo-zy** [reply]
Can you tell me how you guys are currently using http proxies, I think I can fix this,

**Reply by RafaelLinux** [reply]
Question is for users or for devs?

**Reply by AlexanderGabriel** [reply]
We have access to different networks (not just internet) via different proxys so we use Proxy-Auto-Config (https://de.wikipedia.org/wiki/Proxy_Auto-Config) in our Browsers to control which Proxy is used. So this would be gold standard ;)
We use Squid running on Port 3128 with BASIC Authentication for Internet Access. Many other apps have a dialoge like "use system proxy" which respects OS-Settings or have fields for proxyhostname, port, user and password or read env-vars like http_proxy, https_proxy, no_proxy. which would be okay, too or just put proxy settings into some kind of ini file or registry would be okay, too so we could push this to our clients via software distribution system or GPO (we are working on Windows)
Squid is configured to allow http_connect method to target port 443 to the internet. Systems like Jitsi do a websocket this way to establish connection to videobridges if users are behind proxys.
We do not have any external dns-resolving.
If you need any help to test something or settings for some test-proxy in your environment, i will try to help you.
If you need funding. Tell me how much. i will be needing a few days to check with my boss. and maybe other users join to fund.
are you part of rustdesk-project? if not: will your PR be accepted or would it be better to ask @rustdesk devs for this to garantee the feature will be in upstream?

**Reply by yuluo-zy** [reply]
Sorry, I'm not a member of the Rustdesk project, I just need this and want to try it, I've tried overloading UDP using the HTTP protocol to simply try it,

**Reply by rustdesk** [reply]
We accept all PR, we will review it strictly before merging it.

**Reply by yuluo-zy** [reply]
Hello, I have set up an squid  http proxy environment locally. @AlexanderGabriel  The relevant configuration is as follows：

```
acl localnet src 0.0.0.0/0
http_port 3128
maximum_object_size_in_memory 64 KB
cache_mem 1 GB
max_filedescriptors 1048576
acl Safe_ports port 3128
acl CONNECT method CONNECT
http_access allow localnet Safe_ports
http_access allow localnet CONNECT Safe_ports
```
 
 Currently, the software runs locally without any problems.
 For the test of https tunnel proxy, I used the https tunnel proxy on the following public website https://proxy-tools.com/， There is no problem so far. They may also use Squid to build https tunnel services.
 
![image](https://github.com/rustdesk/rustdesk/assets/33770335/0794d990-2995-4438-b58e-90ce3b1e2c35)
![image](https://github.com/rustdesk/rustdesk/assets/33770335/146651cc-49d9-4d29-87e2-8f1cdbd53c5b)
![image](https://github.com/rustdesk/rustdesk/assets/33770335/49dfa95d-93b1-443a-bb1a-753caffd8da5)

 @tboernert In your environment, what did you use to build the proxy? I think I can continue to test it after building it locally.

**Reply by tboernert** [reply]
yes we use also the squid proxy. normally with browser autokonfiguration. so there is no konfiguration needed on the clients. the client looks for an dns entry with wpad and there are we running a small lighttpd that will give the proxy autoconfig script back to the client. this will normally done from operating system.
https://developer.mozilla.org/en-US/docs/Web/HTTP/Proxy_servers_and_tunneling/Proxy_Auto-Configuration_PAC_file

**Reply by AlexanderGabriel** [reply]
same here (wpad) but it would be ok, if there was a config file where we can set proxy. where is this configured?
can "acl Safe_ports port 3128" be "acl Safe_ports port 443" and rust server run on port 443?
can i test your code?

**Reply by michag86** [reply]
And it would be nice if this would work behind a reverse proxy. So that rustdesk listens on a high port and something like traefik or nginx can Run as reverse proxy in front of rustdesk.

**Reply by yuluo-zy** [reply]
The functionality for manually configuring an HTTP proxy is now supported. I will submit a Pull Request for merging over the weekend. Afterwards, we can conduct some manual HTTP tests.

**Reply by rustdesk** [reply]
How about https?

**Reply by rustdesk** [reply]
Both tested for controlled side and controlling side?

**Reply by yuluo-zy** [reply]
Yes, I have already tested it. I used an HTTPS proxy service from an open website and was able to connect and control normally. （public website [https://proxy-tools.com/](https://proxy-tools.com/%EF%BC%8C)）

**Comment by rustdesk** [comment]
@yuluo-zy Thanks, you can follow our socks implementation, start from here, https://github.com/rustdesk/rustdesk/blob/master/libs/hbb_common/src/socket_client.rs.

https://github.com/rustdesk/rustdesk/pulls?q=is%3Apr+is%3Aclosed+socks

**Comment by rustdesk** [comment]
The problem is UDP.

**Comment by yuluo-zy** [comment]
> The problem is UDP.

Senat on udp?

I will review the project code and propose related implementation.

**Reply by rustdesk** [reply]
The good news is UDP is not mandatory any more in RustDesk.

**Reply by AlexanderGabriel** [reply]
no more udp sounds good!

**Reply by rustdesk** [reply]
The bad news is the OSS version self-hosting server still requires UDP.

**Reply by sao-coding** [reply]
When can I build a self-built server in the OSS version without using UDP?

**Comment by borondics** [comment]
Hello, do you have a timeline on when we could expect the http(s) proxy addition? Thanks a lot in advance!!!

**Comment by rustdesk** [comment]
This is a good example if you want a feature request get high priority.
https://github.com/rustdesk/rustdesk/discussions/7437

**Reply by tboernert** [reply]
can you please tell us the price for this feature ? thank you very much.

**Reply by rustdesk** [reply]
@yuluo-zy you can give a price if you wanna do this project as a contributor.

**Reply by yuluo-zy** [reply]
I think I can complete this function. It is expected to be completed in one to two months. Can the price be 850 US dollars?

**Reply by tboernert** [reply]
oh sorry i read that to late. price is ok. you can start. what will do you need from my site ?
thank you very much !

**Reply by rustdesk** [reply]
Please fund here, https://github.com/rustdesk/rustdesk/issues/7487, alternatively, you can also use paypal, info@rustdesk.com

**Reply by yuluo-zy** [reply]
Thank you very much. It’s not needed yet. When I finish the development, I will contact you as soon as possible and test it in your environment to ensure the effectiveness and correctness of the function. @tboernert

**Comment by AlexanderGabriel** [comment]
So tell us how much to fund this and i will talk to my boss. I think it would be better to get a quote and then an invoice because we are public administration and to get a permission for a donation is a really long process

**Reply by rustdesk** [reply]
It depends on you, and if it can attract contributor. We do not do this (we do not change our path), but we try to advocate it and attract contributor.

**Reply by rustdesk** [reply]
I think we can give you invoice.

**Reply by tboernert** [reply]
sounds good

**Reply by rustdesk** [reply]
https://github.com/rustdesk/rustdesk/discussions/5687#discussioncomment-8884339

**Comment by rustdesk** [comment]
What stop you using the SOCKS proxy which has better performance?

**Reply by tboernert** [reply]
SOCKS is no solutions to setup that on the customer firewalls. http proxy ist availiable.

**Reply by AlexanderGabriel** [reply]
I understand the idea behind your question but i really don't want to start a discussion wheater to use socks or http. our system is to big, to "just" use another proxy, too much dependencies, our proxy is well tested and works, integratet to our security-conepts, etc. changing is not an option. and http proxy protocol is really old, good documented and well testet. i don't know any other organization like us that uses socks. hope this answer okay for you

**Comment by rustdesk** [comment]
SOCKS proxy can be configured to 443 port also.

**Reply by michag86** [reply]
The problem is that in many restricted Enterprise networks there is only a http proxy that allows HTTPS on port 443.  So everything hast Tonne tunneled through Port 443. Maybe something like a reverse proxy can help but IT has to pass the http proxy.

**Reply by rustdesk** [reply]
Thanks, I understand now. @michag86 @AlexanderGabriel

**Reply by tboernert** [reply]
Thank you!
can you please give us a price for the implementation?
I have also many customers who i can test this ...

**Reply by rustdesk** [reply]
https://github.com/rustdesk/rustdesk/discussions/5687#discussioncomment-8884339

**Comment by yuluo-zy** [comment]
httpThe http function has been merged, you can use the nightly version to experience it

**Comment by rustdesk** [comment, marked as answer]
Close as finished, https://twitter.com/rustdesk/status/1783468475480191022

`http/https` proxy does not support UDP, so it does not work for the free oss rustdesk server as contolled side. Our public server and Pro version support work without UDP.

**Comment by AlexanderGabriel** [comment]
I can test next week. Our firewall team first has to unlock target Port of rustdesk servers in proxy and firewall.

**Comment by michag86** [comment]
Sound great but there is still the problem with the high ports.

Would be nice if there is a solution to put the rustdesk server behind a reverse proxy like traefik to get it working with only 443/TCP.

**Comment by rouben** [comment]
Perhaps there's a way to do this with [HTTP/3 aka QUIC](https://en.wikipedia.org/wiki/HTTP/3), which uses UDP? Also, AFAIK Traefik supports arbitrary (non-HTTP) TCP/UDP protocol proxying, much like HAProxy, no?

**Comment by michag86** [comment]
Squid proxy for example does not support rfc  9114	HTTP/3 and also does not support QUIC.
https://wiki.squid-cache.org/StandardsCompliance
